python/ray/experimental/data/impl/compute.py

```diff
@@ -1,54 +1,80 @@
-from typing import TypeVar, List, Any
+from typing import TypeVar, Iterable, Any
 
 import ray
 from ray.experimental.data.impl.block import Block, ObjectRef
+from ray.experimental.data.impl.block_list import BlockList, BlockMetadata
 from ray.experimental.data.impl.progress_bar import ProgressBar
 
 T = TypeVar("T")
 U = TypeVar("U")
 
 
 class ComputePool:
-    def apply(self, fn: Any, blocks: List[Block[T]]) -> List[ObjectRef[Block]]:
+    def apply(self, fn: Any,
+              blocks: Iterable[Block[T]]) -> Iterable[ObjectRef[Block]]:
         raise NotImplementedError
 
 
 class TaskPool(ComputePool):
     def apply(self, fn: Any, remote_args: dict,
-              blocks: List[Block[T]]) -> List[ObjectRef[Block]]:
+              blocks: BlockList[Any]) -> BlockList[Any]:
         map_bar = ProgressBar("Map Progress", total=len(blocks))
 
-        if remote_args:
-            fn = ray.remote(**remote_args)(fn)
-        else:
-            fn = ray.remote(fn)
-        blocks = [fn.remote(b) for b in blocks]
+        kwargs = remote_args.copy()
+        kwargs["num_returns"] = 2
 
-        map_bar.block_until_complete(blocks)
-        return blocks
+        @ray.remote(**kwargs)
+        def wrapped_fn(block: Block, meta: BlockMetadata):
+            new_block = fn(block)
+            new_meta = BlockMetadata(
+                num_rows=new_block.num_rows(),
+                size_bytes=new_block.size_bytes(),
+                schema=new_block.schema(),
+                input_files=meta.input_files)
+            return new_block, new_meta
+
+        refs = [
+            wrapped_fn.remote(b, m)
+            for b, m in zip(blocks, blocks.get_metadata())
+        ]
+        new_blocks = [r[0] for r in refs]
+        new_metadata = [r[1] for r in refs]
```

Comment:
Nit: Use `zip()` idiom.
```suggestion
        new_blocks, new_metadata = zip(*refs)
```

Reply:
I don't think this makes it more readable.